torch/csrc/autograd/input_buffer.cpp

```diff
@@ -184,15 +184,15 @@ void InputBuffer::add(
       if (on_consumer && !on_producer) {
         // (3a)
         opt_accumulate_stream = opt_consumer_stream;
-        opt_sync_stream = guard.getDefaultStream(opt_consumer_stream->device());
+        opt_sync_stream = guard.getNewStream(opt_consumer_stream->device());
```

Comment:
I'm not really sure about creating a new stream every time here.
In particular, backends without stream caching or limited number of stream might suffer in terms of perf.
And more importantly, I think it will make profiler trace very hard to read.

Not using the default stream sounds like a fair ask but then I think we should continue to have only one stream.
A solution here would be to have autograd lazily get one stream per-deviceidx that will be used for all accumulations.

Reply:
> A solution here would be to have autograd lazily get one stream per-deviceidx that will be used for all accumulations.
Do you mean that we hold a stream per-deviceidx. and we can switch to one of them in different cases ? And I have a question, why we need a default stream here? for my naïve understanding. We just need to make accumulate happens on producer or consumer stream, and consumer stream wait for producer stream. thanks.

Reply:
> Do you mean that we hold a stream per-deviceidx

Yes 

> We just need to make accumulate happens on producer or consumer stream, and consumer stream wait for producer stream. thanks.

It happens in the (admittedly weird) case where the device idx of both the producer and consumer streams are not the same as the device idx of the Tensor. And we want to accumulate on the device of the Tensor. So we need a stream on that device idx where we don't have one.
That's why we went with the default stream as any stream on that device would do.